File: backend/app/core/upgrades/abilities/triple_dash.py
from ..upgrade_implementation import UpgradeImplementation, UpgradeAppliers


class TripleDash(UpgradeImplementation):
    id = 'triple_dash'
    name = 'Triple Dash'
    description = 'Store 3 dashes.'
    rarity = 'legendary'
    type = 'ability'
    cost = 40

    def can_apply(self, state: dict) -> bool:
        return True

    def apply(self, appliers: UpgradeAppliers) -> None:
        appliers.effects.register_effect('triple_dash', {})

    def validate(self, recorded_stats: dict) -> bool:
        return True

File: frontend/src/game/upgrades/abilities/triple_dash.ts
import { Upgrade, type UpgradeDef, type UpgradeContext } from '../Upgrade'
import { RarityID, UpgradeTypeID } from '../../data/ID'

export const TripleDashDef: UpgradeDef = {
  id: "triple_dash",
  name: "Triple Dash",
  description: "Store 3 dashes.",
  rarity: RarityID.Legendary,
  upgradeType: UpgradeTypeID.Ability,
  cost: 40,
  effect: "triple_dash",
  stackable: false,
  dependentOn: [{ ids: ["double_dash"] }],
}

export class TripleDash extends Upgrade {
  onApply(ctx: UpgradeContext): void {
    super.onApply(ctx)
    ctx.player?.setMaxDashCharges(3)
  }
}
